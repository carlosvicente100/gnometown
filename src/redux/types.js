// gnomes
export const fetchRequested = 'FETCH_REQUESTED';
export const fetchRequestedGnomes = 'FETCH_REQUESTED_GNOMES';
export const fetchPendingGnomes = 'FETCH_PENDING_GNOMES';
export const fetchSucceededGnomes = 'FETCH_SUCCEEDED_GNOMES';
export const fetchFailedGnomes = 'FETCH_FAILED_GNOMES';

//filters
export const setAgeFromTo = 'SET_BREWED_FROM_TO';
export const setAgeFrom = 'SET_BREWED_FROM';
export const setAgeTo = 'SET_BREWED_TO';
