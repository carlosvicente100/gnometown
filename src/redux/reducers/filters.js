import { fetchSetJobList, fetchSetName, fetchResetFilters } from '../types'

export const initialState = {
  jobList: [],
  name: ''
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSetJobList:
      return {
        ...state,
        jobList: action.data.jobs
      }
    case fetchSetName:
      return {
        ...state,
        name: action.data.name
      }
    case fetchResetFilters:
      return {
        ...initialState
      }
    default:
      return {
        ...state
      }
  }
}
