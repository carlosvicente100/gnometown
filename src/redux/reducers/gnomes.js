import { fetchPendingGnomes, fetchSucceededGnomes, fetchFailedGnomes } from '../types'

export const initialState = {
  pending: false,
  error: null,
  townName: '',
  gnomes: []
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchPendingGnomes:
      return {
        ...state,
        pending: true
      }
    case fetchSucceededGnomes:
      const townName = Object.keys(action.data)[0]
      return {
        ...state,
        pending: false,
        townName,
        gnomes: action.data[townName]
      }
    case fetchFailedGnomes:
      return {
        ...state,
        pending: false,
        error: true
      }
    default:
      return {
        ...state
      }
  }
}
