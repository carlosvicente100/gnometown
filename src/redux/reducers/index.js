import { combineReducers } from 'redux';
 import gnomesReducer from './gnomes';
import filtersReducer from './filters';

export default combineReducers({
  gnomesReducer,
  filtersReducer,
});
