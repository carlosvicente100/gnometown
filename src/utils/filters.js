export const getJobsByUsers = (users) => {
  //   console.log('getJobsByUsers', users)
  if (users.length > 0) {
    const profesionList = users
      .map((user) => user.professions)
      .reduce((firstArray, secondArray) => [...firstArray, ...secondArray], [])
    return profesionList.filter((job, index) => profesionList.indexOf(job) === index)
  } else {
    return []
  }
}

export const DEFAULT_VALUE = ''
export const JOBS = 'Jobs'

export const makePagination = (itemsFiltered) => {
  return itemsFiltered.reduce((result, value, index) => {
    if (index % 40 !== 0) {
      result[result.length - 1].push(value)
    } else {
      result.push([value])
    }
    return result
  }, [])
}

export const filterByJob = (jobs, users) => {
  let gnomesWithJobs = users.filter((user) => {
    if (jobs.every((job) => user.professions.includes(job))) {
      return user
    } else {
      return false
    }
  })
  return gnomesWithJobs
}
