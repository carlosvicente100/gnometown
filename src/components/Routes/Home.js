import React, { useState } from 'react'
import styled from 'styled-components'
import { useSelector } from 'react-redux'
import PropTypes from 'prop-types'

import { makePagination } from '../../utils'
import Banner from '../Banner/Banner'
import ItemList from '../Gnome/gnomeList'
import Pagination from '../Pagination'

//at the moment, not used
const HomeContainer = styled.div`
  font-family: verdana;
  height: -webkit-fill-available;
  height: 98vh;
  background-color: #2c2c54;
  display: grid;
  grid-template-columns: auto 500px auto;
  row-gap: 20px;
  grid-template-areas:
    ". Filters ."
    "ItemList ItemList ItemList"
    "NoResults NoResults NoResults"
    ". Pagination .";

  div.Filters {
    grid-area: Filters;
  }
  div.ItemList {
    grid-area: ItemList;
  }
  div.NoResults {
    background:#2980b9;
    color:white;
    display: flex;
    justify-content: center;
    align-items: center;
    width 100%;
    height:300px;
    grid-area: NoResults;
  }
  div.Pagination {
    grid-area: Pagination;
  }
`

const Home = () => {
  const [currentPage, setCurrentPage] = useState(0)
  const { gnomes = [], error } = useSelector((state) => state.gnomesReducer)
  const { startDate, endDate } = useSelector((state) => state.filtersReducer)
  let gnomesFiltered = gnomes // []
  //filter
  /* if (gnomes.length > 0) {
        gnomesFiltered = filterByDate(startDate, endDate, items)
      }*/

  let paginatedGnomes = makePagination(gnomesFiltered)

  /*
  const updateFilters = () => {
    setCurrentPage(0)
  }
      */
  console.log('paginatedGnomes', paginatedGnomes)

  return (
    <>
      <Banner isHome={true}></Banner>
      {paginatedGnomes.length > 0 ? (
        <>
          <ItemList items={paginatedGnomes[currentPage]}></ItemList>
          <Pagination
            setCurrentPage={setCurrentPage}
            totalPagination={paginatedGnomes}
            current={currentPage}
          ></Pagination>
        </>
      ) : (
        <div className="NoResults">
          <p>
            {error ? "We're having problems Searching Gnomes, try it in a few minutes" : "no gnomes, don't be exigent "}
          </p>
        </div>
      )}
    </>
  )
}

Home.propTypes = {}

export default Home
