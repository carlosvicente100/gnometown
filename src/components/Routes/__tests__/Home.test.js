import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { render as rtlRender } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { initialState as initialStateGnomes } from '../../../redux/reducers/gnomes'
import { initialState as initialStateFilters } from '../../../redux/reducers/filters'
import reducer from '../../../redux/reducers'

import Home from '../Home'

import GnomeFixtureList from '../../../fixtures/gnomes'

const initialReducerState = {
  gnomesReducer: initialStateGnomes,
  filtersReducer: initialStateFilters
}

function render(ui, { initialState, store = createStore(reducer, initialState), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

describe('Home Route Component', () => {
  test('Home Route Snapshot', () => {
    expect(render(<Home />, { initialReducerState }).baseElement).toMatchSnapshot()
  })

  test('Should Show Current Gnome List lenght - gnomesReducer', () => {
    const gnomeList = GnomeFixtureList[0]['Brastlewark']
    const store = createStore(() => ({
      filtersReducer: {
        jobList: [],
        name: ''
      },
      gnomesReducer: {
        townName: 'Rivendel',
        pending: false,
        error: null,
        gnomes: gnomeList
      }
    }))

    const { getByText, findByText } = render(<Home />, { store })
    expect(getByText(/Current Gnomes:/i)).toBeTruthy()

    expect(findByText(/Current Gnomes: #[3-5]/)).toBeTruthy()
  })
})
