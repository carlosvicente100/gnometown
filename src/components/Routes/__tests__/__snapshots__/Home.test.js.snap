// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Route Component Home Route Snapshot 1`] = `
.c1 {
  position: relative;
  background: white;
  width: 100%;
  height: 72px;
  display: grid;
  grid-template-columns: 15% 35% 35% 15%;
  grid-template-areas: '. m-town-name m-town-name m-filter';
  margin-bottom: 20px;
}

.c1 label {
  text-align: center;
  grid-area: m-town-name;
  font-family: 'Helvetica';
  font-weight: 300;
  font-size: 34px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
}

.c3 {
  display: grid;
  grid-template-columns: auto;
  margin: auto;
  -webkit-column-gap: 20px;
  column-gap: 20px;
  row-gap: 5px;
  margin: 20px;
}

.c3 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  margin: auto;
  margin: 0;
  width: 100%;
}

.c2 {
  margin: auto;
  text-align: center;
  margin-bottom: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 28px;
}

.c2 i {
  margin: 0 20px;
  cursor: pointer;
  border: 1px solid grey;
  border-radius: 10px;
  padding: 5px;
  box-shadow: 1px 1px 5px 0px rgba(50,50,50,0.75);
}

.c2 .open {
  color: #ffbe76;
}

.c2 .active {
  color: tomato;
}

.c0 {
  text-align: center;
  font-family: Helvetica;
}

.c0 .NoResults {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 300px;
  grid-area: NoResults;
}

@media (min-width:688px) {

}

@media (min-width:900px) {

}

@media (min-width:1201px) {
  .c3 {
    row-gap: 20px;
    grid-template-columns: auto auto auto auto;
  }

  .c3 a {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
    margin: auto;
  }
}

@media (min-width:1501px) {
  .c3 {
    grid-template-columns: auto auto auto auto auto;
  }
}

<body>
  <div>
    <div
      class="c0"
      data-testid="Home-page"
    >
      <div
        class="c1"
      >
        <label
          data-testid="banner-town-name"
        >
           citizens
        </label>
      </div>
      <div
        class="c2"
      >
        <i
          class="fas fa-hammer    "
          data-testid="job-filter-button"
        />
        <i
          class="fas fa-user   "
          data-testid="name-filter-button"
        />
      </div>
      <div
        class="filterActive"
        data-testid="filter-container"
      />
      <div
        class="gnomes-length"
      >
        Current Gnomes: 0
      </div>
      <div
        class="c3 GnomeList"
        data-testid="gnome-list-container"
      >
        <div>
          no gnomes founded :(
        </div>
      </div>
    </div>
  </div>
</body>
`;
